drivers/platform/maxim/maxim_uart.c

```diff
@@ -0,0 +1,353 @@
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *  - Neither the name of Analog Devices, Inc. nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include <errno.h>
+#include <stdlib.h>
+#include "maxim_stdio.h"
+#include "mxc_sys.h"
+#include "uart.h"
+#include "maxim_uart.h"
+#include "no-os/uart.h"
+#include "no-os/irq.h"
+#include "no-os/util.h"
+
+/**
+* @brief Callback descriptor that contains a function to be called when an
+* interrupt occurs.
+*/
+static struct callback_desc *cb[2];
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+void UART0_IRQHandler()
+{
+	if(!cb[0])
+		return;
+
+	mxc_uart_regs_t *uart_regs = MXC_UART0;
+	uint8_t	n_int = 0;
+	volatile uint32_t reg_int = uart_regs->int_fl;
+
+	/** Clear all interrupt flags */
+	uart_regs->int_fl = 0x0;
+
+	while(reg_int) {
+		if((reg_int & 1) && uart_regs->int_en && BIT(n_int)) {
+			void *ctx = cb[0]->ctx;
+			void *config = cb[0]->config;
+			cb[0]->callback(ctx, n_int, config);
+		}
+		n_int++;
+		reg_int >>= 1;
+	}
```

Comment:
This doesn't seem to work correctly. I propose:
`while (reg_int) {
  n_int = find_first_set_bit(reg_int);
  if (reg_int & (uart_regs->int_en & BIT(n_int))) {
    //call cb
  }
  reg_int >>= n_int + 1;
}`